### PR TITLE
feat: add strategy support for state archives

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

Make the state-archive process more generic so we can extend in future.  

**Description**

- Make the state-archive process to generic 
- Add feature flag to the cli to select a strategy 

**Steps to test or reproduce**

- Run all tests

### CODE REVIEW

File: packages/beacon-node/src/chain/archiver/archiver.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
+import {Logger} from "@lodestar/utils";
```
Comment (nazarhussain) [inline packages/beacon-node/src/chain/archiver/archiver.ts:1]: This file is actually old `index.ts` file, with only strategy specific changes.

File: packages/beacon-node/src/chain/archiver/interface.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import {CheckpointWithHex} from "@lodestar/fork-choice";
+import {Metrics} from "../../metrics/metrics.js";
+import {RootHex} from "@lodestar/types";
+
+export enum ArchiveMode {
+  Full = "full",
```
Comment (nazarhussain) [inline packages/beacon-node/src/chain/archiver/interface.ts]: If you think of a better name for current archiver strategy we use, please suggest.

File: packages/cli/src/options/beaconNodeOptions/chain.ts [hunk 3]
```diff
@@ -210,6 +212,16 @@ Will double processing times. Use only for debugging purposes.",
     group: "chain",
   },
 
+  "chain.archiveMode": {
+    hidden: false,
+    choices: Object.values(ArchiveMode),
+    description: "Strategy to manage archive states",
+    default: defaultOptions.chain.archiveMode,
+    defaultDescription: String(defaultOptions.chain.archiveMode),
```
Comment (nflaig) [inline packages/cli/src/options/beaconNodeOptions/chain.ts]: `defaultDescription` is not required if you have `default` set

File: packages/cli/src/options/beaconNodeOptions/chain.ts [hunk 4]
```diff
@@ -210,6 +212,16 @@ Will double processing times. Use only for debugging purposes.",
     group: "chain",
   },
 
+  "chain.archiveMode": {
+    hidden: false,
```
Comment (nflaig) [inline packages/cli/src/options/beaconNodeOptions/chain.ts]: this flag should probably be hidden

File: packages/beacon-node/src/chain/archiver/interface.ts (hunk 2 above)
Comment (nflaig) [inline packages/beacon-node/src/chain/archiver/interface.ts]: More of a general question, once we have differential archive mode implemented, will there be any advantage to use our current (full) archiving strategy?

If not, we should probably not make it overly generic / switchable

Comment (nazarhussain) [reply packages/beacon-node/src/chain/archiver/interface.ts]: My idea is to get the differential strategy released as experimental first. As I am a bit skeptical about it's memory implication. So during that time user can choose which strategy to use.

File: packages/cli/src/options/beaconNodeOptions/chain.ts (hunk 4 above)
Comment (nazarhussain) [reply packages/cli/src/options/beaconNodeOptions/chain.ts]: It's a user facing flag not for our testing only. Also it have a default value so exposing would not harm.

Comment (nflaig) [reply packages/cli/src/options/beaconNodeOptions/chain.ts]: exposing it in this PR does not make sense imo, we can always do that later

File: packages/cli/src/options/beaconNodeOptions/chain.ts [hunk 5]
```diff
@@ -210,6 +212,15 @@ Will double processing times. Use only for debugging purposes.",
     group: "chain",
   },
 
+  "chain.stateArchiveMode": {
+    hidden: true,
+    choices: Object.values(StateArchiveMode),
+    description: "Strategy to manage archive states",
```
Comment (twoeths) [inline packages/cli/src/options/beaconNodeOptions/chain.ts]: want to be more clear about the description

```suggestion
    description: `Strategy to manage archive states, only support ${DEFAULT_STATE_ARCHIVE_MODE} at this time`,
```

File: packages/beacon-node/src/chain/archiver/strategies/frequencyStateArchiveStrategy.ts [hunk 6]
```diff
@@ -1,34 +1,23 @@
...
-  archiveStateEpochFrequency: number;
-}
+import {IBeaconDb} from "../../../db/index.js";
+import {IStateRegenerator} from "../../regen/interface.js";
+import {getStateSlotFromBytes} from "../../../util/multifork.js";
+import {serializeState} from "../../serializeState.js";
+import {AllocSource, BufferPool} from "../../../util/bufferPool.js";
+import {Metrics} from "../../../metrics/metrics.js";
+import {StateArchiveStrategy, StatesArchiverOpts} from "../interface.js";
+import {PERSIST_TEMP_STATE_EVERY_EPOCHS} from "../constants.js";
```
Comment (twoeths) [inline packages/beacon-node/src/chain/archiver/strategies/frequencyStateArchiveStrategy.ts]: I prefer to have constants specific to this file's strategy to stay on this file

File: packages/beacon-node/src/chain/archiver/interface.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
+import {CheckpointWithHex} from "@lodestar/fork-choice";
+import {Metrics} from "../../metrics/metrics.js";
+import {RootHex} from "@lodestar/types";
```
Comment (twoeths) [inline packages/beacon-node/src/chain/archiver/interface.ts:3]: the new linter does not enforce import order anymore?

File: packages/beacon-node/src/chain/archiver/interface.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+export enum StateArchiveMode {
+  Frequency = "frequency",
+  // Specify only existing strategy
```
Comment (twoeths) [inline packages/beacon-node/src/chain/archiver/interface.ts]: may want to link the the state diff issue, something like:
// new strategy to be implemented in https://github.com/ChainSafe/lodestar/issue/xxxx

File: packages/beacon-node/src/chain/archiver/archiver.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+          this.chain.emitter.off(ChainEvent.checkpoint, this.onCheckpoint);
+        },
+        {once: true}
+      );
+    }
+  }
+
+  /** Archive latest finalized state */
+  async persistToDisk(): Promise<void> {
+    await this.statesArchiverStrategy.maybeArchiveState(this.chain.forkChoice.getFinalizedCheckpoint());
```
Comment (twoeths) [inline packages/beacon-node/src/chain/archiver/archiver.ts]:
```suggestion
    return this.statesArchiverStrategy.maybeArchiveState(this.chain.forkChoice.getFinalizedCheckpoint());
```

**Review by twoeths — COMMENTED** [review]
the approach lgtm for the 1st round, dropped some minor comments

File: packages/beacon-node/src/chain/archiver/interface.ts (hunk 7 above)
Comment (nazarhussain) [reply packages/beacon-node/src/chain/archiver/interface.ts:3]: There is a caveat related to import order, not yet enabled. https://github.com/ChainSafe/lodestar/issues/7182

**Review by twoeths — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.23.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.23.0) :tada:

**Pull request merged**